### ISSUE TITLE
[Bug] PatchCore ONNX and OpenVINO pred_score mismatch with ckpt inference (anomalib 2.6.0, clean Docker environment)

### ISSUE THREAD

**Issue by ricklina90** [open-edge-platform/anomalib#3726]
### Describe the bug

First of all, I would like to express my sincere gratitude to the anomalib team 
for providing such an excellent anomaly detection library. It has been incredibly 
helpful for our industrial defect inspection work.

I also want to thank the team for the previous assistance provided in the related 
Discussion and Issue. I apologize in advance if any information in this report is 
insufficient or unclear — please feel free to let me know.

Based on the feedback from the previous Discussion #3644 and Issue #3658, 
I conducted additional testing using a clean Docker environment to rule out 
any local environment contamination. Unfortunately, the issue persists.

When using PatchCore with wide_resnet101_2 backbone, the pred_score from both 
the ONNX export and OpenVINO export do not match the pred_score from the ckpt inference.

More critically, the pred_label is also inconsistent — the same NG image is 
correctly identified as anomalous by the ckpt model, but incorrectly classified 
as normal by both the ONNX and OpenVINO models (using the official /255.0 preprocessing method).

An important finding: ONNX and OpenVINO produce almost identical pred_score results 
(difference: 0.0003), which suggests the issue originates in the export pipeline itself, 
not in the inference engine.

This is a serious issue for industrial defect inspection applications, where 
a false negative (NG classified as OK) is unacceptable.

Related links:
- Discussion: https://github.com/open-edge-platform/anomalib/discussions/3644
- Previous Issue: https://github.com/open-edge-platform/anomalib/issues/3658

Test files (model.ckpt / model.onnx / NG_x640.png):
- OneDrive: https://1drv.ms/u/c/91da903f9f216365/IQAEGX6djctSQ6AHxhk9xJ0jAReZWzUiHyHocdRPVxT8NRM?e=k1pv6p

Note: The dataset field above is selected as MVTecAD for form compliance.
The actual dataset used is a custom Folder dataset with the same structure.

### Dataset

MVTecAD

### Model

PatchCore

### Steps to reproduce the behavior

1. Build Docker image using the provided Dockerfile (see Configuration YAML section)

```bash
docker build -t anomalib-trainer:latest .
```

2. Prepare dataset under workspace directory:
output/ABC/datasets/ABC/
├── train/
│ └── ABC_good/ ← normal training images
└── test/
├── ABC_good/ ← normal test images
└── ABC_defect/ ← abnormal test images

3. Run training using train_script.py (see Configuration YAML section):

```bash
docker run --rm --gpus all \
  -v "D:\your\workspace:/workspace" \
  anomalib-trainer:latest \
  /opt/venv/bin/python /workspace/train_script.py
```

4. Place `model.ckpt`, `model.onnx`, and test image `NG_x640.png` in workspace directory.
   (Test files available on OneDrive — see Describe the bug section)

5. Run inference comparison using test_inference.py (see Configuration YAML section):

```bash
docker run --rm --gpus all \
  -v "D:\your\workspace:/workspace" \
  anomalib-trainer:latest \
  /opt/venv/bin/python /workspace/test_inference.py
```

6. Observe the mismatch in pred_score and pred_label (see Logs section):

| Method | pred_score | pred_label |
|--------|-----------|------------|
| ckpt inference | 0.6645 | True (NG) |
| ONNX (Official /255.0) | 0.4667 | **False (OK) ❌** |
| ONNX (ImageNet Normalize) | 1.0000 | True (NG) |
| OpenVINO (Official /255.0) | 0.4823 | **False (OK) ❌** |

Key finding: ONNX and OpenVINO pred_score are almost identical (difference: 0.0003),
confirming the issue is in the export pipeline, not the inference engine.

### OS information

- OS: Windows 11 + Docker Desktop v4.87.0 (WSL2 backend)
- Docker Base Image: nvidia/cuda:12.6.3-cudnn-devel-ubuntu22.04 (Ubuntu 22.04)
- Python version: 3.12 (inside Docker, managed by uv)
- Anomalib version: 2.6.0
- PyTorch version: cu130 (installed via anomalib[cu130])
- CUDA/cuDNN version: CUDA 12.6.3 / cuDNN (Driver version: 13.2)
- GPU models and configuration: NVIDIA GeForce RTX 5090 (32GB)
- CPU: Intel Core i7-14700F
- OpenVINO version: 2026.3.0
- Any other relevant information: Clean Docker environment to eliminate local package contamination. OpenVINO inference tested on CPU only (i7-14700F has no integrated GPU).

### Expected behavior

The pred_score and pred_label from both ONNX and OpenVINO inference should match 
the results from ckpt inference when using the same input image.

Expected:
ckpt     pred_score : 0.6645  pred_label : True (NG)
ONNX     pred_score : ~0.6645 pred_label : True (NG)
OpenVINO pred_score : ~0.6645 pred_label : True (NG)

### Screenshots

Not applicable. Results are shown in text format in the Steps section.

### Pip/GitHub

pip

### What version/branch did you use?

anomalib : 2.6.0

### Configuration YAML

```yaml
No YAML configuration used.
Training and inference are done entirely via Python API.


======== Dockerfile ========

FROM nvidia/cuda:12.6.3-cudnn-devel-ubuntu22.04
ENV DEBIAN_FRONTEND=noninteractive
ENV PYTHONIOENCODING=utf-8
ENV NO_COLOR=1
ENV TERM=dumb
ENV PATH="/root/.local/bin:/opt/venv/bin:$PATH"
RUN apt-get update && apt-get install -y \
    curl git wget build-essential libssl-dev zlib1g-dev \
    libgl1-mesa-glx libglib2.0-0 && rm -rf /var/lib/apt/lists/*
RUN curl -LsSf https://astral.sh/uv/install.sh | sh
RUN uv python install 3.12
RUN uv venv /opt/venv --python 3.12
RUN uv pip install --python /opt/venv/bin/python "anomalib[cu130]"
RUN uv pip install --python /opt/venv/bin/python onnx onnxruntime-gpu
RUN uv pip install --python /opt/venv/bin/python openvino
WORKDIR /workspace
CMD ["/opt/venv/bin/python", "--version"]


======== train_script.py ========

import sys, io, torch, logging, warnings, os
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8')
from anomalib.data import Folder
from anomalib.models import Patchcore
from anomalib.engine import Engine
from anomalib.deploy import ExportType
os.environ['LIGHTNING_RANK_ZERO_LOG_LEVEL'] = 'ERROR'
logging.getLogger('lightning.pytorch').setLevel(logging.ERROR)
warnings.filterwarnings('ignore')
torch.set_float32_matmul_precision('high')
if __name__ == '__main__':
    datamodule = Folder(
        name='ABC',
        root=r'/workspace/output/ABC/datasets/ABC',
        normal_dir='train/ABC_good',
        normal_test_dir='test/ABC_good',
        abnormal_dir='test/ABC_defect',
        test_split_mode='from_dir',
        val_split_mode='same_as_test',
        train_batch_size=8,
        eval_batch_size=8,
        num_workers=0,
    )
    model = Patchcore(
        backbone='wide_resnet101_2',
        coreset_sampling_ratio=0.3,
        num_neighbors=2,
    )
    engine = Engine(
        max_epochs=1,
        default_root_dir=r'/workspace/output/ABC/results',
    )
    print('=== Start Training ===')
    engine.fit(datamodule=datamodule, model=model)
    print('=== Start Testing ===')
    results = engine.test(datamodule=datamodule, model=model)
    if results:
        metrics = results[0] if isinstance(results, list) else results
        if isinstance(metrics, dict):
            auroc   = metrics.get('image_AUROC',   metrics.get('image_auroc',   0.0))
            f1score = metrics.get('image_F1Score', metrics.get('image_f1score', 0.0))
            print(f'image_AUROC:{auroc:.4f}')
            print(f'image_F1Score:{f1score:.4f}')
    print('Exporting ONNX...')
    engine.export(
        model=model,
        export_type=ExportType.ONNX,
        export_root=r'/workspace/output/ABC/results',
    )
    print('Export ONNX complete.')
    print('Exporting OpenVINO...')
    engine.export(
        model=model,
        export_type=ExportType.OPENVINO,
        export_root=r'/workspace/output/ABC/results',
    )
    print('Export OpenVINO complete.')


======== test_inference.py ========

import sys, io, warnings, numpy as np
from pathlib import Path
from PIL import Image
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8')
warnings.filterwarnings('ignore')
TEST_DIR    = Path('/workspace')
CKPT_PATH   = TEST_DIR / 'model.ckpt'
ONNX_PATH   = TEST_DIR / 'model.onnx'
OV_XML_PATH = TEST_DIR / 'model.xml'
IMAGE_PATH  = TEST_DIR / 'NG_x640.png'
from anomalib.models import Patchcore
from anomalib.engine import Engine
from anomalib.data import PredictDataset
from torch.utils.data import DataLoader
import onnxruntime as ort
from torchvision import transforms
from openvino import Core

model      = Patchcore(backbone='wide_resnet101_2')
engine     = Engine()
dataset    = PredictDataset(path=IMAGE_PATH)
dataloader = DataLoader(dataset, batch_size=1, collate_fn=dataset.collate_fn)
ckpt_result = engine.predict(
    model=model, dataloaders=dataloader, ckpt_path=str(CKPT_PATH))
ckpt_score = None
if ckpt_result:
    pred       = ckpt_result[0] if isinstance(ckpt_result, list) else ckpt_result
    ckpt_score = pred.pred_score.item()
    print(f'ckpt pred_score : {ckpt_score:.4f}')
    print(f'ckpt pred_label : {pred.pred_label.item()}')

# Method 1: ONNX Official /255.0
session    = ort.InferenceSession(str(ONNX_PATH),
    providers=['CUDAExecutionProvider', 'CPUExecutionProvider'])
input_meta = session.get_inputs()[0]
image      = Image.open(IMAGE_PATH).convert('RGB').resize((256, 256))
image_np   = np.asarray(image)
image_chw  = np.transpose(image_np, (2, 0, 1))[None, ...]
tensor1    = image_chw.astype(np.float32) / 255.0
outputs1   = session.run(None, {input_meta.name: tensor1})
names      = [o.name for o in session.get_outputs()]
for name, out in zip(names, outputs1):
    if name in ('pred_score', 'pred_label'):
        print(f'ONNX (Official /255.0) {name} : {out[0]}')

# Method 2: ONNX ImageNet Normalize
transform  = transforms.Compose([
    transforms.Resize((256, 256)),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485,0.456,0.406], std=[0.229,0.224,0.225])])
tensor2    = transform(Image.open(IMAGE_PATH).convert('RGB')).unsqueeze(0).numpy()
outputs2   = session.run(names, {input_meta.name: tensor2})
for name, out in zip(names, outputs2):
    if name in ('pred_score', 'pred_label'):
        print(f'ONNX (ImageNet Normalize) {name} : {out[0]}')

# Method 3: OpenVINO Official /255.0
if OV_XML_PATH.exists():
    ie       = Core()
    model_ov = ie.read_model(str(OV_XML_PATH))
    compiled = ie.compile_model(model_ov, 'CPU')
    input_layer = compiled.input(0)
    image_ov    = Image.open(IMAGE_PATH).convert('RGB').resize((256, 256))
    image_np    = np.asarray(image_ov)
    image_chw   = np.transpose(image_np, (2, 0, 1))[None, ...]
    input_data  = image_chw.astype(np.float32) / 255.0
    result_ov   = compiled({input_layer.any_name: input_data})
    for out in compiled.outputs:
        name    = out.any_name
        out_arr = np.asarray(result_ov[out])
        if 'pred_score' in name:
            print(f'OpenVINO (Official /255.0) pred_score : {out_arr.flatten()[0]:.4f}')
        if 'pred_label' in name:
            print(f'OpenVINO (Official /255.0) pred_label : {bool(out_arr.flatten()[0])}')
```

### Logs

```shell
=== Path Check ===
ckpt     : /workspace/model.ckpt
onnx     : /workspace/model.onnx
openvino : /workspace/model.xml
image    : /workspace/NG_x640.png

=== ckpt Inference ===
pred_score : 0.6645
pred_label : True

=== ONNX Inference - Method 1: Official Preprocessing (/255.0) ===
  Input  name  : input
  Input  type  : tensor(float)
  Input  shape : (1, 3, 256, 256)
  Output count : 4
  Output[0] name=pred_score  shape=(1,)          min=0.466680  max=0.466680
  Output[1] name=pred_label  shape=(1,)          min=0.000000  max=0.000000
  Output[2] name=anomaly_map shape=(1,1,256,256) min=0.401687  max=0.591924
  Output[3] name=pred_mask   shape=(1,1,256,256) min=0.000000  max=1.000000
pred_score : 0.4667
pred_label : False

=== ONNX Inference - Method 2: ImageNet Normalize Preprocessing ===
pred_score : 1.0000
pred_label : True

=== OpenVINO Inference (Official /255.0) ===
  Input  name  : input
  Input  shape : [?,3,?,?]
  Output names : ['pred_score', 'pred_label', 'anomaly_map', 'pred_mask']
  Output name=pred_score  shape=(1,) min=0.482311  max=0.482311
pred_score : 0.4823
  Output name=pred_label  shape=(1,) min=0.000000  max=0.000000
pred_label : False

=== Comparison ===
ckpt       pred_score                      : 0.6645
ONNX       pred_score (Official   /255.0)  : 0.4667
ONNX       pred_score (ImageNet Normalize) : 1.0000
OpenVINO   pred_score (Official   /255.0)  : 0.4823
Difference (ONNX     Official   /255.0)    : 0.1978  Match : False
Difference (ONNX     ImageNet Normalize)   : 0.3355  Match : False
Difference (OpenVINO Official   /255.0)    : 0.1732  Match : False

Key Finding:
ONNX and OpenVINO pred_score are almost identical (difference: 0.0003)
This confirms the issue originates in the export pipeline, not the inference engine.
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by ricklina90** [comment]
## Additional Testing: OpenVINO Export Validation

I have conducted additional testing to further investigate the issue.
The OneDrive link in the bug description has been updated with the latest files,
including the OpenVINO model files (model.xml, model.bin) and updated scripts.

### Updated OneDrive Contents
- Dockerfile (updated: added openvino)
- train_script.py (updated: added OpenVINO export)
- test_inference.py (updated: added OpenVINO comparison)
- model.ckpt
- model.onnx
- model.xml (OpenVINO model structure)
- model.bin (OpenVINO model weights)
- NG_x640.png (test image)

OneDrive: https://1drv.ms/u/c/91da903f9f216365/IQAEGX6djctSQ6AHxhk9xJ0jAReZWzUiHyHocdRPVxT8NRM?e=k1pv6p

### OpenVINO Test Results

| Method | pred_score | pred_label |
|--------|-----------|------------|
| ckpt inference | 0.6645 | True (NG) |
| ONNX (Official /255.0) | 0.4667 | **False (OK) ❌** |
| ONNX (ImageNet Normalize) | 1.0000 | True (NG) |
| OpenVINO (Official /255.0) | 0.4823 | **False (OK) ❌** |

### Key Finding

ONNX and OpenVINO pred_score are almost identical (difference: 0.0003),
which strongly suggests the issue originates in the **export pipeline itself**,
not in the inference engine (ONNX Runtime or OpenVINO Runtime).

Since OpenVINO is converted from ONNX internally by anomalib,
both formats inherit the same issue from the export process.

This further confirms that the post-processing or score normalization
is not correctly included during the export process.

Thank you for looking into this. Any guidance would be greatly appreciated.
